internal/pkg/otel/extension/elasticdiagnostics/extension.go

```diff
@@ -0,0 +1,239 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package elasticdiagnostics
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"runtime/pprof"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/confmap"
+	"go.uber.org/zap"
+	"go.yaml.in/yaml/v3"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/proto"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent/internal/pkg/diagnostics"
+	"github.com/elastic/elastic-agent/pkg/ipc"
+)
+
+var (
+	_ component.Component = &diagnosticsExtension{}
+)
+
+type diagHook struct {
+	description string
+	filename    string
+	contentType string
+	hook        func() []byte
+}
+
+type diagnosticsExtension struct {
+	listener net.Listener
+	server   *http.Server
+	logger   *zap.Logger
+	logp     *logp.Logger
+
+	diagnosticsConfig *Config
+	collectorConfig   *confmap.Conf
+	componentHooks    map[string][]*diagHook
+	globalHooks       map[string]*diagHook
+
+	hooksMtx sync.Mutex
+	confgMtx sync.Mutex
+}
+
+func (d *diagnosticsExtension) Start(ctx context.Context, host component.Host) error {
+	var err error
+
+	d.logp, err = logp.NewZapLogger(d.logger)
+	if err != nil {
+		// NewZapLogger always returns nil error, so this shouldn't happen.
+		return fmt.Errorf("failed to create logp.Logger from zap logger: %w", err)
+	}
+
+	d.registerGlobalDiagnostics()
+
+	d.listener, err = ipc.CreateListener(d.logp, d.diagnosticsConfig.Endpoint)
+	if err != nil {
+		return fmt.Errorf("error creating listener: %w", err)
+	}
+
+	mux := http.NewServeMux()
+	mux.Handle("/diagnostics", d)
+
+	d.server = &http.Server{
+		Handler:           mux,
+		ReadHeaderTimeout: 30 * time.Second,
+	}
+	go func() {
+		if err := d.server.Serve(d.listener); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			d.logger.Error("HTTP server error", zap.Error(err))
+		}
+	}()
+	d.logger.Info("Diagnostics extension started", zap.String("address", d.listener.Addr().String()))
+	return nil
+}
+
+func (d *diagnosticsExtension) Shutdown(ctx context.Context) error {
+	if d.server == nil {
+		return nil
+	}
+	if err := d.server.Shutdown(ctx); err != nil {
+		return err
+	}
+	ipc.CleanupListener(d.logp, d.diagnosticsConfig.Endpoint)
+	return nil
+}
+
+func (d *diagnosticsExtension) registerGlobalDiagnostics() {
+	d.globalHooks["collector_config"] = &diagHook{
+		description: "full collector configuration",
+		filename:    "edot/otel-merged.yaml",
```

Comment:
```suggestion
		filename:    "edot/otel-merged-actual.yaml",
```

We use the expected vs actual naming elsewhere and I think it would apply here too, the one at the root is the expected from agent and the one here is what the collector was last (actually) run.